barretenberg/cpp/src/barretenberg/stdlib/primitives/group/cycle_group.cpp

```diff
@@ -0,0 +1,1324 @@
+        auto lhs = cycle_group::from_constant_witness(context, get_value());
+        return lhs.unconditional_add(other);
+    }
+    if (!lhs_constant && rhs_constant) {
+        auto rhs = cycle_group::from_constant_witness(context, other.get_value());
+        return unconditional_add(rhs);
+    }
+
+    const auto p1 = get_value();
+    const auto p2 = other.get_value();
+    AffineElement p3(Element(p1) + Element(p2));
+    if (lhs_constant && rhs_constant) {
+        return cycle_group(p3);
+    }
+    field_t r_x(witness_t(context, p3.x));
+    field_t r_y(witness_t(context, p3.y));
+    cycle_group result(r_x, r_y, false);
+
+    proof_system::ecc_add_gate_<FF> add_gate{
+        .x1 = x.get_witness_index(),
+        .y1 = y.get_witness_index(),
+        .x2 = other.x.get_witness_index(),
+        .y2 = other.y.get_witness_index(),
+        .x3 = result.x.get_witness_index(),
+        .y3 = result.y.get_witness_index(),
+        .endomorphism_coefficient = 1,
+        .sign_coefficient = 1,
+    };
+    context->create_ecc_add_gate(add_gate);
+
+    return result;
+}
+
+/**
+ * @brief will evaluate ECC point subtraction over `*this` and `other`.
+ *        Incomplete addition formula edge cases are *NOT* checked!
+ *        Only use this method if you know the x-coordinates of the operands cannot collide
+ *
+ * @tparam Composer
+ * @param other
+ * @return cycle_group<Composer>
+ */
+template <typename Composer>
+cycle_group<Composer> cycle_group<Composer>::unconditional_subtract(const cycle_group& other) const
+{
+    if constexpr (!IS_ULTRA) {
+        return unconditional_add(-other);
+    } else {
+        auto context = get_context(other);
+
+        const bool lhs_constant = is_constant();
+        const bool rhs_constant = other.is_constant();
+
+        if (lhs_constant && !rhs_constant) {
+            auto lhs = cycle_group<Composer>::from_constant_witness(context, get_value());
+            return lhs.unconditional_subtract(other);
+        }
+        if (!lhs_constant && rhs_constant) {
+            auto rhs = cycle_group<Composer>::from_constant_witness(context, other.get_value());
+            return unconditional_subtract(rhs);
+        }
+        auto p1 = get_value();
+        auto p2 = other.get_value();
+        AffineElement p3(Element(p1) - Element(p2));
+        if (lhs_constant && rhs_constant) {
+            return cycle_group(p3);
+        }
+        field_t r_x(witness_t(context, p3.x));
+        field_t r_y(witness_t(context, p3.y));
+        cycle_group result(r_x, r_y, false);
+
+        proof_system::ecc_add_gate_<FF> add_gate{
+            .x1 = x.get_witness_index(),
+            .y1 = y.get_witness_index(),
+            .x2 = other.x.get_witness_index(),
+            .y2 = other.y.get_witness_index(),
+            .x3 = result.x.get_witness_index(),
+            .y3 = result.y.get_witness_index(),
+            .endomorphism_coefficient = 1,
+            .sign_coefficient = -1,
```

Comment:
Ah the main difference between this and unconditional_add for UltraPlonk is the sign_coefficient and the fact that we are doing point subtraction `AffineElement p3(Element(p1) - Element(p2))`

Not blocking for this PR, in a later PR we can find a way to reduce this duplication of code

Reply:
aye good spot, we can reduce code duplication by merging these implementations. Will do so in next PR